File: app/src/main/java/com/sjn/stamp/ui/activity/DrawerActivity.java
package com.sjn.stamp.ui.activity;

import android.content.res.Configuration;
import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.google.android.gms.cast.framework.CastButtonFactory;
import com.google.android.gms.cast.framework.CastContext;
import com.mikepenz.materialdrawer.AccountHeader;
import com.mikepenz.materialdrawer.AccountHeaderBuilder;
import com.mikepenz.materialdrawer.Drawer;
import com.mikepenz.materialdrawer.DrawerBuilder;
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem;
import com.mikepenz.materialdrawer.model.interfaces.Nameable;
import com.sjn.stamp.R;
import com.sjn.stamp.ui.DrawerMenu;
import com.sjn.stamp.utils.AnalyticsHelper;
import com.sjn.stamp.utils.LogHelper;

public abstract class DrawerActivity extends AppCompatActivity implements FragmentManager.OnBackStackChangedListener {

    private static final String TAG = LogHelper.makeLogTag(DrawerActivity.class);
    public static final String FRAGMENT_TAG = "fragment_container";

    protected Toolbar mToolbar;

    protected Drawer mDrawer;
    private ActionBarDrawerToggle mDrawerToggle;
    protected AccountHeader mAccountHeader;

    private long mNextDrawerMenu;
    protected long mSelectingDrawerMenu;
    private boolean mToolbarInitialized;

    public boolean onOptionsItemSelected(int itemId) {
        return false;
    }

    protected void changeFragmentByDrawer(long selectedDrawerMenu) {
        DrawerMenu drawerMenu = DrawerMenu.of(selectedDrawerMenu);
        if (drawerMenu == null) {
            return;
        }
        navigateToBrowser(drawerMenu.getFragment(), false, selectedDrawerMenu);
        setToolbarTitle(null);
    }

    public void setToolbarTitle(CharSequence title) {
        if (title == null) {
            title = getSelectingDrawerName();
        }
        setTitle(title);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.main, menu);
        try {
            CastButtonFactory.setUpMediaRouteButton(getApplicationContext(),
                    menu, R.id.media_route_menu_item);
        } catch (RuntimeException e) {
            e.printStackTrace();
        }
        return true;
    }

    @Override
    public void onBackStackChanged() {
        updateDrawerToggleState();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        LogHelper.d(TAG, "Activity onCreate");
        try {
            CastContext.getSharedInstance(this);
        } catch (RuntimeException e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (!mToolbarInitialized) {
            throw new IllegalStateException("You must run super.initializeToolbar at " +
                    "the end of your onCreate method");
        }
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        if (mDrawerToggle != null) {
            mDrawerToggle.syncState();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        // Whenever the fragment back stack changes, we may need to update the
        // action bar toggle: only top level screens show the hamburger-like icon, inner
        // screens - either Activities or fragments - show the "Up" icon instead.
        getSupportFragmentManager().addOnBackStackChangedListener(this);
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if (mDrawerToggle != null) {
            mDrawerToggle.onConfigurationChanged(newConfig);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        getSupportFragmentManager().removeOnBackStackChangedListener(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    final public boolean onOptionsItemSelected(MenuItem item) {
        if (mDrawerToggle != null && mDrawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        // If not handled by drawerToggle, home needs to be handled by returning to previous
        if (item != null && item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        } else if (item != null) {
            if (!onOptionsItemSelected(item.getItemId())) {
                return super.onOptionsItemSelected(item);
            }
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        // If the drawer is open, back will close it
        if (mDrawer != null && mDrawer.isDrawerOpen()) {
            mDrawer.closeDrawer();
            return;
        }
        // Otherwise, it may return to the previous fragment stack
        FragmentManager fragmentManager = getSupportFragmentManager();
        if (fragmentManager.getBackStackEntryCount() > 0) {
            fragmentManager.popBackStack();
        } else {
            moveTaskToBack(true);
            // Lastly, it will rely on the system behavior for back
            // super.onBackPressed();
        }
    }

    @Override
    public void setTitle(CharSequence title) {
        super.setTitle(title);
        mToolbar.setTitle(title);
    }

    @Override
    public void setTitle(int titleId) {
        super.setTitle(titleId);
        mToolbar.setTitle(titleId);
    }

    public void navigateToBrowser(Fragment fragment, boolean addToBackStack, long selection) {
        navigateToBrowser(fragment, addToBackStack);
        mDrawer.setSelection(selection);
    }

    public void navigateToBrowser(Fragment fragment, boolean addToBackStack) {
        if (!addToBackStack) {
            getSupportFragmentManager().popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
        }
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.container, fragment, FRAGMENT_TAG);
        if (addToBackStack) {
            // If this is not the top level media (root), we add it to the fragment back stack,
            // so that actionbar toggle and Back will work appropriately:
            //transaction.setCustomAnimations(R.anim.fade_out, R.anim.fade_in);
            transaction.addToBackStack(null);
        }
        transaction.commit();
        AppBarLayout appBarLayout = (AppBarLayout) findViewById(R.id.app_bar);
        appBarLayout.setExpanded(true, true);
    }

    protected void initializeToolbar() {
        mToolbar = (Toolbar) findViewById(R.id.toolbar);
        if (mToolbar == null) {
            throw new IllegalStateException("Layout is required to include a Toolbar with id " +
                    "'toolbar'");
        }
        mToolbar.inflateMenu(R.menu.main);
        setSupportActionBar(mToolbar);
        mSelectingDrawerMenu = getCurrentMenuId();
        createDrawer();
        mToolbarInitialized = true;
    }

    protected long getCurrentMenuId() {
        if (mDrawer != null) {
            return mDrawer.getCurrentSelection();
        }
        return DrawerMenu.first().getMenuId();
    }

    private void updateDrawerToggleState() {
        if (mDrawerToggle == null) {
            return;
        }
        boolean isRoot = getSupportFragmentManager().getBackStackEntryCount() == 0;
        mDrawerToggle.setDrawerIndicatorEnabled(isRoot);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowHomeEnabled(!isRoot);
            getSupportActionBar().setDisplayHomeAsUpEnabled(!isRoot);
            getSupportActionBar().setHomeButtonEnabled(!isRoot);
        }
        if (isRoot) {
            mDrawerToggle.syncState();
        }
    }

    private String getSelectingDrawerName() {
        if (mDrawer != null && mDrawer.getDrawerItem(mDrawer.getCurrentSelection()) instanceof Nameable) {
            return ((Nameable) mDrawer.getDrawerItem(mDrawer.getCurrentSelection())).getName().getText(this);
        }
        return getString(R.string.app_name);
    }

    private void createDrawer() {
        createAccountHeader();
        mDrawer = new DrawerBuilder().withActivity(this)
                .withAccountHeader(mAccountHeader)
                .withToolbar(mToolbar)
                .inflateMenu(R.menu.drawer)
                .withSelectedItem(mSelectingDrawerMenu)
                .withOnDrawerNavigationListener(new Drawer.OnDrawerNavigationListener() {
                    @Override
                    public boolean onNavigationClickListener(View clickedView) {
                        //this method is only called if the Arrow icon is shown. The hamburger is automatically managed by the MaterialDrawer
                        //if the back arrow is shown. close the activity
                        onBackPressed();
                        //return true if we have consumed the event
                        return true;
                    }
                })
                .withOnDrawerItemClickListener(new Drawer.OnDrawerItemClickListener() {
                    @Override
                    public boolean onItemClick(View view, int position, IDrawerItem drawerItem) {
                        AnalyticsHelper.trackScreen(DrawerActivity.this, DrawerMenu.of(drawerItem.getIdentifier()));
                        setNextDrawerMenu(drawerItem);
                        return false;
                    }
                })
                .withOnDrawerListener(new Drawer.OnDrawerListener() {
                    @Override
                    public void onDrawerOpened(View view) {
                    }

                    @Override
                    public void onDrawerClosed(View view) {
                        if (mSelectingDrawerMenu == mNextDrawerMenu) {
                            return;
                        }
                        mSelectingDrawerMenu = mNextDrawerMenu;
                        changeFragmentByDrawer(mNextDrawerMenu);
                    }

                    @Override
                    public void onDrawerSlide(View view, float v) {
                    }
                })
                .build();

        mDrawerToggle = mDrawer.getActionBarDrawerToggle();
    }

    private void setNextDrawerMenu(IDrawerItem drawerItem) {
        mNextDrawerMenu = drawerItem.getIdentifier();
    }

    private void createAccountHeader() {
        mAccountHeader = new AccountHeaderBuilder()
                .withSelectionListEnabledForSingleProfile(false)
                .withActivity(this)
                .withHeaderBackground(R.drawable.drawer_header).build();
    }
}

File: app/src/main/java/com/sjn/stamp/ui/activity/DrawerActivity.kt
package com.sjn.stamp.ui.activity

import android.app.Activity
import android.content.res.Configuration
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.design.widget.CollapsingToolbarLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.content.ContextCompat
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import com.github.sjnyag.forceanimateappbarlayout.ForceAnimateAppBarLayout
import com.google.android.gms.cast.framework.CastButtonFactory
import com.google.android.gms.cast.framework.CastContext
import com.sjn.stamp.R
import com.sjn.stamp.ui.DrawerMenu
import com.sjn.stamp.ui.fragment.media.PagerFragment
import com.sjn.stamp.utils.CompatibleHelper
import com.sjn.stamp.utils.DrawerHelper
import com.sjn.stamp.utils.LogHelper
import com.sjn.stamp.utils.ToolbarColorizeHelper
import io.multimoon.colorful.CAppCompatActivity


abstract class DrawerActivity : CAppCompatActivity(), FragmentManager.OnBackStackChangedListener {

    private var toolbar: Toolbar? = null
    protected var drawer: DrawerHelper.Drawer? = null

    abstract fun onOptionsItemSelected(itemId: Int): Boolean

    open fun setToolbarTitle(title: CharSequence?) = setTitle(title ?: drawer?.selectingDrawerName
    ?: "")

    open fun navigateToBrowser(fragment: Fragment, addToBackStack: Boolean, sharedElements: List<Pair<String, View>> = emptyList()) {
        if (!addToBackStack) {
            supportFragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        }
        supportFragmentManager.beginTransaction().apply {
            replace(R.id.container, fragment, FRAGMENT_TAG)
            if (addToBackStack) {
                addToBackStack(null)
            }
            sharedElements.forEach { pair -> addSharedElement(pair.second, pair.first) }
        }.commit()
    }

    fun animateAppbar(title: String?, updateImageView: (activity: Activity, imageView: ImageView) -> Unit) {
        findViewById<Toolbar>(R.id.toolbar)?.apply {
            setTitle(title)
        }
        findViewById<CollapsingToolbarLayout>(R.id.toolbar_layout)?.apply {
            isTitleEnabled = true
            setTitle(title)
        }
        findViewById<ImageView>(R.id.app_bar_image)?.also { imageView ->
            imageView.visibility = View.VISIBLE
            updateImageView(this, imageView)
        }
        findViewById<ForceAnimateAppBarLayout>(R.id.app_bar).run {
            animatedExpand()
        }
    }

    fun updateAppbar() {
        val hasShadow = supportFragmentManager.findFragmentByTag(FRAGMENT_TAG) !is PagerFragment
        findViewById<AppBarLayout>(R.id.app_bar).run {
            if (CompatibleHelper.hasLollipop()) {
                elevation = if (hasShadow) 8F else 0F
            }
        }
        findViewById<CollapsingToolbarLayout>(R.id.toolbar_layout)?.apply {
            isTitleEnabled = false
        }
        findViewById<ImageView>(R.id.app_bar_image)?.apply {
            visibility = View.GONE
        }
    }

    protected fun initializeToolbar() {
        toolbar = findViewById<View>(R.id.toolbar) as Toolbar?
        toolbar?.let {
            it.inflateMenu(R.menu.main)
            setSupportActionBar(it)
            drawer = DrawerHelper.Drawer(this, it, object : DrawerHelper.Drawer.Listener {
                override fun changeFragmentByDrawer(menu: Long) {
                    val drawerMenu = DrawerMenu.of(menu) ?: return
                    navigateToBrowser(drawerMenu.fragment, false)
                    setToolbarTitle(null)
                }
            })
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.main, menu)
        try {
            CastButtonFactory.setUpMediaRouteButton(applicationContext, menu, R.id.media_route_menu_item)
        } catch (e: RuntimeException) {
            e.printStackTrace()
        }
        return true
    }

    override fun onBackStackChanged() {
        drawer?.updateDrawerToggleState()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        LogHelper.d(TAG, "Activity onCreate")
        try {
            CastContext.getSharedInstance(this)
        } catch (e: RuntimeException) {
            e.printStackTrace()
        }
    }

    override fun onResume() {
        super.onResume()
        supportFragmentManager.addOnBackStackChangedListener(this)
    }

    override fun onPause() {
        super.onPause()
        supportFragmentManager.removeOnBackStackChangedListener(this)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        drawer?.sync()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawer?.onConfigurationChanged(newConfig)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        drawer?.let {
            if (it.onOptionItemSelected(item)) {
                return true
            }
        }
        if (item != null && item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        } else if (item != null) {
            if (!onOptionsItemSelected(item.itemId)) {
                return super.onOptionsItemSelected(item)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        drawer?.let {
            if (it.closeDrawer()) {
                return
            }
        }
        val fragmentManager = supportFragmentManager
        if (fragmentManager.backStackEntryCount > 0) {
            fragmentManager.popBackStack()
        } else {
            moveTaskToBack(true)
        }
    }

    override fun setTitle(title: CharSequence) {
        super.setTitle(title)
        toolbar?.title = title
    }

    override fun setTitle(titleId: Int) {
        super.setTitle(titleId)
        toolbar?.setTitle(titleId)
    }

    companion object {

        private val TAG = LogHelper.makeLogTag(DrawerActivity::class.java)
        const val FRAGMENT_TAG = "fragment_container"
    }
}
